apps/web/pages/signup.tsx

```diff
@@ -34,43 +43,130 @@ type FormValues = z.infer<typeof signupSchema>;
+      if (!debouncedUsername) {
+        setPremium(false);
+        setUsernameTaken(false);
+        return;
+      }
+      fetchUsername(debouncedUsername).then(({ data }) => {
+        setPremium(data.premium);
+        setUsernameTaken(!data.available);
+      });
+    }
+    checkUsername();
+  }, [debouncedUsername, setPremium, setUsernameTaken]);
 
-const getOrgUsernameFromEmail = (email: string, autoAcceptEmailDomain: string) => {
-  const [emailUser, emailDomain] = email.split("@");
-  const username =
-    emailDomain === autoAcceptEmailDomain
-      ? slugify(emailUser)
-      : slugify(`${emailUser}-${emailDomain.split(".")[0]}`);
+  return (
+    <div>
+      <TextField
+        {...props}
+        {...register("username")}
+        data-testid="signup-usernamefield"
+        addOnFilled={false}
+      />
+      {!formState.isSubmitting && (
+        <div className="text-gray text-default flex items-center text-sm">
+          <p className="flex items-center text-sm ">
+            {usernameTaken ? (
+              <div className="text-error">
+                <Info className="mr-1 inline-block h-4 w-4" />
+                {t("already_in_use_error")}
+              </div>
+            ) : premium ? (
+              <div data-testid="premium-username-warning">
+                <StarIcon className="mr-1 inline-block h-4 w-4" />
+                {t("premium_username", {
+                  price: getPremiumPlanPriceValue(),
+                })}
+              </div>
+            ) : null}
+          </p>
+        </div>
+      )}
+    </div>
+  );
+}
 
-  return username;
-};
+const checkValidEmail = (email: string) => z.string().email().safeParse(email).success;
 
 function addOrUpdateQueryParam(url: string, key: string, value: string) {
   const separator = url.includes("?") ? "&" : "?";
   const param = `${key}=${encodeURIComponent(value)}`;
   return `${url}${separator}${param}`;
 }
 
-export default function Signup({ prepopulateFormValues, token, orgSlug, orgAutoAcceptEmail }: SignupProps) {
+export default function Signup({ prepopulateFormValues, token, orgSlug }: SignupProps) {
+  const [premiumUsername, setPremiumUsername] = useState(false);
+  const [usernameTaken, setUsernameTaken] = useState(false);
+
   const searchParams = useSearchParams();
   const telemetry = useTelemetry();
   const { t, i18n } = useLocale();
+  const router = useRouter();
   const flags = useFlagMap();
-  const methods = useForm<FormValues>({
-    mode: "onChange",
+  const formMethods = useForm<FormValues>({
     resolver: zodResolver(signupSchema),
-    defaultValues: prepopulateFormValues,
+    defaultValues: prepopulateFormValues satisfies FormValues,
   });
   const {
     register,
-    formState: { errors, isSubmitting },
-  } = methods;
+    watch,
+    formState: { isSubmitting, errors },
+  } = formMethods;
 
-  const handleErrors = async (resp: Response) => {
+  const handleErrorsAndStripe = async (resp: Response) => {
```

Comment:
Stolen from website approach - premium username throws a 402 if a premium username is required. 

We could probably handle this better but it outside the scope of this PR